components/raftstore/src/coprocessor/split_observer.rs

```diff
@@ -1,85 +1,83 @@
+        let mut slice = key.as_slice();
+        let strip_len = match bytes::decode_bytes(&mut slice, false) {
+            // slice points to the remaining unparsable part of the key
+            Ok(_) => slice.len(),
+            // it's a raw key so no need to strip
+            Err(_) => 0,
         };
-
-        let key = encode_bytes(&key);
-        match util::check_key_in_region_exclusive(&key, region) {
-            Ok(()) => Ok(key),
-            Err(_) => Err(format!(
-                "key {} should be in ({}, {})",
-                log_wrappers::Value::key(&key),
-                log_wrappers::Value::key(&region.get_start_key()),
-                log_wrappers::Value::key(&region.get_end_key()),
-            )),
-        }
+        key.truncate(key.len() - strip_len);
+        key
     }
 
     fn on_split(
         &self,
         ctx: &mut ObserverContext<'_>,
         splits: &mut Vec<SplitRequest>,
-    ) -> Result<()> {
+    ) -> Result<(), String> {
         let (mut i, mut j) = (0, 0);
         let mut last_valid_key: Option<Vec<u8>> = None;
-        let region_id = ctx.region().get_id();
+        let region = ctx.region();
+        let region_id = region.get_id();
         while i < splits.len() {
             let k = i;
             i += 1;
             {
                 let split = &mut splits[k];
-                let key = split.take_split_key();
-                match self.adjust_key(ctx.region(), key) {
-                    Ok(key) => {
-                        if last_valid_key.as_ref().map_or(false, |k| *k >= key) {
-                            warn!(
-                                "key is not larger than previous, skip.";
-                                "region_id" => region_id,
-                                "key" => log_wrappers::Value::key(&key),
-                                "previous" => log_wrappers::Value::key(last_valid_key.as_ref().unwrap()),
-                                "index" => k,
-                            );
-                            continue;
-                        }
-                        last_valid_key = Some(key.clone());
-                        split.set_split_key(key)
-                    }
-                    Err(e) => {
-                        warn!(
-                            "invalid key, skip";
-                            "region_id" => region_id,
-                            "index" => k,
-                            "err" => ?e,
-                        );
-                        continue;
-                    }
+                let key = self.may_strip_timestamp(split.take_split_key());
+
+                if key.is_empty() {
+                    warn!(
+                        "key is empty";
+                        "region_id" => region_id,
+                        "index" => k,
+                    );
+                    continue;
                 }
+
+                if let Err(Error::KeyNotInRegion(..)) =
```

Comment:
Can you make it a minimum change by keeping `adjust_keys` and move theses checks into `adjust_keys`?

Reply:
I think the new name more precisely described what it does. Until 1 year ago, there are other adjustment logic, but then removed, so only strip timestamp now. And I think the check is more suitable in the `on_split()`.

Reply:
It's harder to read than before. I don't think the name matters, it can be named as `adjust_keys` or `check_keys`, the point is put those validations into a function and let outer function to do error handling.

Reply:
done.